rfcs/0013-ai-policies.md

```diff
@@ -0,0 +1,491 @@
+- New users: Can adopt Kuadrant to secure and manage LLM workloads from the start.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- How error would be reported to the users.
+- Corner cases are dissected by example.
+
+The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.
+
+
+## `LLMTokenRateLimitPolicy`
+
+- Add a new resource, `LLMTokenRateLimitPolicy`, to be managed by the Kuadrant operator.
+- Either:
+  - Extend our existing `wasm-shim` to optionally amend the existing `actionSet` to optionally call both the guard filter and the token parsing filter implementation.
+  - Create a new `ext_proc` gRPC service for parsing OpenAI-style and usage metrics and adding these as well-known dynamic metadata, for use by Limitador
+- Extend the wasm-chim and `RateLimitPolicy` to give a means to specify an increment (currently, [hard-coded](https://github.com/Kuadrant/wasm-shim/blob/main/src/service/rate_limit.rs#L18) to `1`)
+- Alter the wasm-shim's `actionSets` actions to inject (where appropriate) additional steps for both guards and token usage metrics parsing. The order of actions matters here, as usage metrics are flushed as part of the body of LLM responses (either complete responses, or when streamed). Some additional notes on our existing filters, including our"internal to WASM" http filter chain, in this thread: https://kubernetes.slack.com/archives/C05J0D0V525/p1744098001098719. A flow diagram below attempts to capture this flow at a high level.
+- Look at ways to support a "two-phase" approach to rate-limiting: a standard, normal rate limit enforcement prior to hitting the model server (responding early if limited), and (if not limited) one after based on one to increment counters by a custom increment (after parsing usage metrics)
+
+### Sequence Diagrams
+
+Below are some sequence diagrams that attempt to capture some of these flows and interactions at a high level.
+
+
+#### Sequence Diagram: Token rate limiting and auth
+
+```mermaid
+sequenceDiagram
+  participant C as Client
+  participant HR as HTTP Route
+  participant E as Envoy Proxy
+  participant A as Authorino
+  participant WL as WASM/EnvoyFilter (Rate Limit & Token Usage)
+  participant L as Limitador
+  participant MS as KServe Model Server
+  participant RPM as Response Processor (Usage Metrics Parser)
+
+  %% initial req
+  C->>HR: Incoming chat/completion request
+  HR->>E: Forward request
+
+  %% JWT authentication check via Authorino
+  E->>A: Validate JWT token & group access
```

Comment:
Minor nit here, but that's `WL` here, not `E`. Fundamentally, `WL` & `E` aren't really distinguishable from a user perspective. But in this particular context, I'd leave `E` out and "just" use `WL`, as that's where we can add/change behaviors as needed.